File: ADBKit/Sources/ADBKit/Features/RowSelection.swift
import Foundation

/// Multi-selection over an ordered list of log rows — the model behind ⌘-click,
/// ⇧-click and drag-select in the streaming feeds. Pure, because the part that
/// goes subtly wrong is never the highlight: it is where the *anchor* sits after
/// each gesture, which decides what the next range spans.
///
/// The order is passed in per call rather than held, because the feed's order is
/// a filtered, reversible, ring-trimmed view that changes under the selection —
/// see `retain(in:)`.
public struct RowSelection<ID: Hashable & Sendable>: Sendable, Equatable {
    public private(set) var ids: Set<ID> = []
    /// The row a range extends from: the last row picked deliberately (a plain
    /// or ⌘ click, or where a drag began), never one swept over by a range.
    public private(set) var anchor: ID?

    public init() {}

    public var isEmpty: Bool { ids.isEmpty }
    public var count: Int { ids.count }
    public func contains(_ id: ID) -> Bool { ids.contains(id) }

    /// A plain click: this row alone.
    public mutating func replace(with id: ID) {
        ids = [id]
        anchor = id
    }

    /// ⌘-click: add or drop one row and leave the rest alone. The row becomes
    /// the anchor either way — the next ⇧-click spans from what was last
    /// touched, which is what makes ⌘-then-⇧ predictable.
    public mutating func toggle(_ id: ID) {
        if ids.contains(id) {
            ids.remove(id)
        } else {
            ids.insert(id)
        }
        anchor = id
    }

    /// ⇧-click: everything from the anchor to `id` in display order. `additive`
    /// keeps what was already selected; otherwise the range replaces it. With no
    /// anchor there is nothing to span, so this behaves like the plain click.
    ///
    /// The anchor deliberately stays put: shift-clicking further down, then
    /// back up, re-spans from the same start instead of ratcheting.
    public mutating func extend(to id: ID, in order: [ID], additive: Bool = false) {
        // A row the feed doesn't have is a click on nothing — leave the
        // selection alone rather than inventing a range around it.
        guard let end = order.firstIndex(of: id) else { return }
        // The anchor, on the other hand, can vanish under a live feed (trimmed,
        // filtered, cleared). Then there is no span, and this is a plain click.
        guard let anchor, let start = order.firstIndex(of: anchor) else {
            if additive { toggle(id) } else { replace(with: id) }
            return
        }
        apply(span: order[min(start, end)...max(start, end)], additive: additive)
    }

    /// A drag: the rows between where the pointer went down and where it is now.
    /// Called repeatedly as the pointer moves, so it always recomputes the span
    /// from the drag's own start — dragging back shrinks the selection again.
    public mutating func select(from start: ID, to end: ID, in order: [ID], additive: Bool = false) {
        guard let first = order.firstIndex(of: start), let last = order.firstIndex(of: end) else { return }
        apply(span: order[min(first, last)...max(first, last)], additive: additive)
        anchor = start
    }

    public mutating func clear() {
        ids = []
        anchor = nil
    }

    /// Drop rows that are no longer in the feed — cleared, filtered out, or
    /// trimmed by the ring buffer. Without this the count and any copy would
    /// include events the reader cannot see.
    public mutating func retain(in order: [ID]) {
        let live = Set(order)
        guard !ids.isSubset(of: live) || anchor.map({ !live.contains($0) }) == true else { return }
        ids.formIntersection(live)
        if let anchor, !live.contains(anchor) { self.anchor = nil }
    }

    /// The selected rows in display order — what a copy has to produce, since a
    /// `Set` has none.
    public func ordered(in order: [ID]) -> [ID] {
        order.filter(ids.contains)
    }

    private mutating func apply(span: ArraySlice<ID>, additive: Bool) {
        ids = additive ? ids.union(span) : Set(span)
    }
}

File: desktop/src/lib/row-selection.ts
/**
 * Picking rows out of a feed — ADBKit's `RowSelection`, ported.
 *
 * **The part that goes wrong is never the highlight, it is the anchor.** The
 * anchor is the row a range extends *from*: the last one picked deliberately,
 * never one swept over by a range. Where it sits after each gesture decides
 * what the next ⇧-click spans, and getting that wrong shows up as a selection
 * that ratchets outward instead of re-spanning, which is the kind of bug you
 * only notice on the fourth click.
 *
 * Immutable where the Mac's is a mutating struct, because this one lives in
 * React state: every operation answers a new value, and the identity check is
 * what lets a feed skip re-rendering rows that did not change.
 */

export interface RowSelection<ID> {
  ids: ReadonlySet<ID>
  /** Where a range extends from, or null when there is nothing to span. */
  anchor: ID | null
}

export function emptySelection<ID>(): RowSelection<ID> {
  return { ids: new Set<ID>(), anchor: null }
}

export function isEmpty<ID>(selection: RowSelection<ID>): boolean {
  return selection.ids.size === 0
}

/** A plain click: this row alone. */
export function replace<ID>(id: ID): RowSelection<ID> {
  return { ids: new Set([id]), anchor: id }
}

/**
 * ⌘-click: add or drop one row and leave the rest alone.
 *
 * The row becomes the anchor either way — the next ⇧-click spans from what was
 * last touched, which is what makes ⌘-then-⇧ predictable.
 */
export function toggle<ID>(selection: RowSelection<ID>, id: ID): RowSelection<ID> {
  const ids = new Set(selection.ids)
  if (ids.has(id)) ids.delete(id)
  else ids.add(id)
  return { ids, anchor: id }
}

/**
 * ⇧-click: everything from the anchor to `id` in display order.
 *
 * `additive` keeps what was already selected; otherwise the range replaces it.
 * The anchor deliberately stays put, so shift-clicking further down and then
 * back up re-spans from the same start rather than ratcheting.
 */
export function extend<ID>(
  selection: RowSelection<ID>,
  id: ID,
  order: readonly ID[],
  additive = false,
): RowSelection<ID> {
  // A row the feed does not have is a click on nothing — leave the selection
  // alone rather than inventing a range around it.
  const end = order.indexOf(id)
  if (end === -1) return selection
  // The anchor, on the other hand, can vanish under a live feed: trimmed,
  // filtered out, cleared. Then there is no span, and this is a plain click.
  const start = selection.anchor === null ? -1 : order.indexOf(selection.anchor)
  if (start === -1) return additive ? toggle(selection, id) : replace(id)
  return {
    ids: spanned(selection, order, start, end, additive),
    anchor: selection.anchor,
  }
}

/**
 * A drag: the rows between where the pointer went down and where it is now.
 *
 * Called on every move, so it recomputes the span from the drag's own start —
 * dragging back up shrinks the selection again instead of leaving a trail.
 */
export function selectRange<ID>(
  selection: RowSelection<ID>,
  from: ID,
  to: ID,
  order: readonly ID[],
  additive = false,
): RowSelection<ID> {
  const first = order.indexOf(from)
  const last = order.indexOf(to)
  if (first === -1 || last === -1) return selection
  return { ids: spanned(selection, order, first, last, additive), anchor: from }
}

/**
 * Drop rows the feed no longer has — cleared, filtered out, or trimmed by the
 * ring buffer.
 *
 * Without this the count and any copy would include events the reader cannot
 * see. Returns the same value when nothing changed, so a feed re-rendering at
 * 60 rows a second does not get a new selection object each time.
 */
export function retain<ID>(selection: RowSelection<ID>, order: readonly ID[]): RowSelection<ID> {
  const live = new Set(order)
  const stale = [...selection.ids].some((id) => !live.has(id))
  const lostAnchor = selection.anchor !== null && !live.has(selection.anchor)
  if (!stale && !lostAnchor) return selection
  return {
    ids: new Set([...selection.ids].filter((id) => live.has(id))),
    anchor: lostAnchor ? null : selection.anchor,
  }
}

/** The selected rows in display order — what a copy needs, since a set has none. */
export function ordered<ID>(selection: RowSelection<ID>, order: readonly ID[]): ID[] {
  return order.filter((id) => selection.ids.has(id))
}

function spanned<ID>(
  selection: RowSelection<ID>,
  order: readonly ID[],
  a: number,
  b: number,
  additive: boolean,
): Set<ID> {
  const span = order.slice(Math.min(a, b), Math.max(a, b) + 1)
  return additive ? new Set([...selection.ids, ...span]) : new Set(span)
}
